gh-90562: Improve zero argument support for `super()` in dataclasses when `slots=True`

Bobronium opened the pull request:
- Original PR: https://github.com/python/cpython/pull/124455

Added two tests for two cases that weren't covered in original PR:
- Using custom descriptors
- Wrapping function before making it a property

<!-- gh-issue-number: gh-90562 -->
* Issue: gh-90562
<!-- /gh-issue-number -->

Bobronium committed:
Improve zero argument support for `super()` in dataclasses

ericvsmith commented on Lib/dataclasses.py:
```diff
@@ -1235,13 +1230,36 @@ def _update_func_cell_for__class__(f, oldcls, newcls):
     return False
 
 
+def _find_inner_functions(obj, _seen=None, _depth=0):
```
Why the leading underscores on `_seen` and `_depth`?

Bobronium replied:
This should indicate that these arguments are only used within the function (in recursive calls) and not outside of the function body.

sobolevn replied:
I think that the whole function is private, so it can only be used inside CPython by us :)

Bobronium replied:
Well, you've got a point. Removed the underscores.

ericvsmith commented on Lib/dataclasses.py:
```diff
@@ -1235,13 +1230,36 @@ def _update_func_cell_for__class__(f, oldcls, newcls):
+    _seen.add(id(obj))
+
+    _depth += 1
+    if _depth > 2:
```
Why are depth 1 and 2 special? What is this test (and `_depth` in general) trying to accomplish? Please add comments explaining what's going on here, it's not at all clear from the code.

Bobronium replied:
Added an explanation and few test cases for this behaviour. Caught two bugs in the process \o/

ericvsmith commented on Lib/dataclasses.py:
```diff
@@ -1326,10 +1347,27 @@ def _add_slots(cls, is_frozen, weakref_slot, defined_fields):
+        ):
+            # we don't want to inspect custom descriptors just yet
+            # there's still a chance we'll encounter a pure function
+            # or a property
```
Please add a comment explaining why encountering a pure function or property means we don't have to inspect custom descriptors.

Bobronium replied:
Hmm, there's already an explanation above:
https://github.com/python/cpython/blob/3b65acf8d66e7e42ad25746976b55dc0bd7d71d8/Lib/dataclasses.py#L1336-L1339

Would you like me to add clarification for this case specifically?

ericvsmith replied:
Ah, you're right. Maybe change the comment to "... fallback to inspecting custom descriptors if no pure function or property is found".

Bobronium replied:
Makes sense. I've added clarification in both places to be extra clear.

Bobronium committed:
Fix NEWS.d formatting

Bobronium committed:
Add explanation for _depth use and tests

Bobronium committed:
Fix incorrect variable reference

ericvsmith commented on Lib/dataclasses.py:
```diff
@@ -1326,10 +1347,27 @@ def _add_slots(cls, is_frozen, weakref_slot, defined_fields):
+                # was wrapped before it became property
+                f = inspect.unwrap(f)
+                if _update_func_cell_for__class__(f, cls, newcls):
+                    break
```
Doesn't this `break` only break out of the inner loop, but you want it to break out of the outer loop? This is the reason I used the `or` clauses in the original. Well that and I didn't want to create the tuple for the inner loop to iterate over, but that's probably an unneeded micro optimization.

Bobronium replied:
Good catch. 

One way to fix this would to use this monstrosity: 
```suggestion
                    break
            else:
                continue
            break
```

Regarding `or` clause, I figured it would be slightly faster than the whole iteration thing, but since there's now unwrapping here, went with the loop for sake of less code duplication.

ericvsmith replied:
Yeah, not a lot of great options here. I think I'd leave the `or` version and add the 3 calls to `unwrap`. Which I realize might make the lines too long, but I think it's better than the above "monstrosity".

Bobronium replied:
Just had another idea how we can handle this: https://github.com/python/cpython/pull/124692/commits/4941a3fdb2d1f23c55b8d45968a78dc2e074ebd1.

ericvsmith replied:
Good one!

Bobronium committed:
Correctly unwrap descriptor function

Bobronium committed:
Add test with using partial as custom wrapper

Bobronium committed:
Rephrase comment explaining _depth argument

Bobronium committed:
Add clarification for function search priority

Bobronium committed:
Ues consistent grammar throughout comments

Bobronium committed:
Fix breaking out from outer loop when handling property

Bobronium committed:
Use generator expression with any() instead of classic loop

Bobronium committed:
Merge branch 'main' into issue-90562-dataclasses-zero-argument-super-fix

sobolevn commented on Lib/dataclasses.py:
```diff
@@ -1235,13 +1230,38 @@ def _update_func_cell_for__class__(f, oldcls, newcls):
+    _depth += 1
+    # Normally just inspection of a descriptor object itself should be enough,
+    # and we should encounter the function as its attribute,
+    # but in case function was wrapped (e.g. functools.partition was used),
```
```suggestion
    # Normally just an inspection of a descriptor object itself should be enough,
    # and we should encounter the function as its attribute,
    # but in case function was wrapped (e.g. functools.partial was used),
```

sobolevn commented on Lib/dataclasses.py:
```diff
@@ -1235,13 +1230,38 @@ def _update_func_cell_for__class__(f, oldcls, newcls):
+        return None
+
+    for attr in dir(obj):
+        value = getattr(obj, attr, None)
```
This case is always complex, since `getattr` can evaluate properties and other objects. There are two ways of hanling this case:
1. Allow all exceptions to raise
2. Hide all exceptions from user here

I am not quite sure which one is better here.

Bobronium replied:
Thanks for bringing this to my attention! 

After some hacking, I've decided to go with the third option: removing any possibility of user defined code execution.

sobolevn commented on Lib/test/test_dataclasses/__init__.py:
```diff
@@ -13,6 +13,8 @@
 import traceback
 import unittest
+from functools import partial
+from functools import update_wrapper
```
```suggestion
from functools import partial, update_wrapper
```

sobolevn reviewed (commented):
Great find, thanks a lot for your work! 👍

Bobronium committed:
Apply suggestions from code review

Co-authored-by: sobolevn <mail@sobolevn.me>

Bobronium committed:
Remove leading underscores from seen and depth args

Bobronium committed:
Prevent user-defined code execution during attribute scanning

carljm commented on Lib/dataclasses.py:
```diff
@@ -1222,11 +1222,6 @@ def _get_slots(cls):
 
 
 def _update_func_cell_for__class__(f, oldcls, newcls):
-    # Returns True if we update a cell, else False.
```
I think this is still the case, so we should keep this comment (or turn it into a docstring.)

carljm reviewed (commented):
This looks OK to me; I'll let @ericvsmith give it final approval/merge.

I don't love that we've opened the door to an infinite fractal of complexity here; at some point it would probably be better to bite the bullet and introduce a `__classcell__` attribute on classes that allows reliably finding this cell in O(1) time.

Bobronium committed:
Use inspect.getattr_static instead of _safe_get_attributes

This uses builtins.dir, which will trigger custom __getattibute__ if any, and will trigger __get__ on __dict__ descriptor.

Bobronium committed:
Put back the comment in _update_func_cell_for__class__

Bobronium committed:
Rely on inspect.getmembers_static()
